### DISCUSSION TITLE
CORS error even when --web.cors.origin flag is passed

### DISCUSSION THREAD

**Discussion by raj-suvariya** [prometheus/prometheus#12071] [Q&A] [answered]
### What did you do?

I am trying to hit Prometheus API but was getting CORS error, searched over internet and found the flag --web.cors.origin can be used to configure CORS. So I added that and now my kubectl describe pod lists this command in Prometheus container

```
    Command:
      /bin/sh
      -c
      sed -e '5s/[()]//g' -e 's/SHARD//g'  -e 's/$NAMESPACE/'"$NAMESPACE"'/g;s/$POD_NAME/'"$POD_NAME"'/g;s/$()/'$(SHARD)'/g' /etc/prometheus/config/prometheus.yml > /etc/prometheus/config_out/prometheus.yml && /bin/prometheus --web.enable-admin-api --web.enable-lifecycle --config.file=/etc/prometheus/config_out/prometheus.yml --storage.tsdb.path=/data/prometheus --storage.tsdb.retention.time=365d --web.read-timeout=5m --web.max-connections=512 --storage.remote.read-concurrent-limit=10 --web.cors.origin=".*"
```

Still when I am trying to hit this from UI we are building it is throwing CORS error.

### What did you expect to see?

CORS error should not occur and Prometheus pod should return the data. This is working when hit form postman

### What did you see instead? Under which circumstances?

I am getting CORS error even when flag to configure cors is passed

### System information

Linux 5.10.0-0.bpo.3-cloud-amd64 x86_64

### Prometheus version

```text
prometheus, version 2.27.1 (branch: HEAD, revision: db7f0bcec27bd8aeebad6b08ac849516efa9ae02)
  build user:       root@fd804fbd4f25
  build date:       20210518-14:17:54
  go version:       go1.16.4
  platform:         linux/amd64
```

### Prometheus configuration file

```yaml
Not needed, let me know if anyone wants specific values to be looked at
```

**Comment by roidelapluie** [comment]
Could you please provide us with more details regarding the error message you are receiving from the UI? Also, please check if the CORS headers are visible in the web dev tools of the browser.

**Comment by raj-suvariya** [comment]
<img width="1515" alt="Screenshot 2023-03-06 at 7 17 20 PM" src="https://user-images.githubusercontent.com/127106399/223128407-a6da5ea0-029c-4ff9-93ed-8fde3ff571a7.png">
<img width="1201" alt="Screenshot 2023-03-06 at 7 18 42 PM" src="https://user-images.githubusercontent.com/127106399/223128418-9f206960-7aa6-468f-bf4c-d84dde195eab.png">

@roidelapluie attaching screenshots for you to look over. API doesn't return any error code it just says CORS error. Also, I don't see any header related to CORS

**Comment by raj-suvariya** [comment]
Also these are server startup logs

```
level=info ts=2023-03-06T13:09:31.370Z caller=main.go:426 msg="Starting Prometheus" version="(version=2.27.1, branch=HEAD, revision=db7f0bcec27bd8aeebad6b08ac849516efa9ae02)"
level=info ts=2023-03-06T13:09:31.370Z caller=main.go:431 build_context="(go=go1.16.4, user=root@fd804fbd4f25, date=20210518-14:17:54)"
level=info ts=2023-03-06T13:09:31.371Z caller=main.go:432 host_details="(Linux 5.10.0-0.bpo.3-cloud-amd64 #1 SMP Debian 5.10.13-1~bpo10+1 (2021-02-11) x86_64 raj-test-hyd-monitor
-0 (none))"
level=info ts=2023-03-06T13:09:31.371Z caller=main.go:433 fd_limits="(soft=1048576, hard=1048576)"
level=info ts=2023-03-06T13:09:31.371Z caller=main.go:434 vm_limits="(soft=unlimited, hard=unlimited)"
level=info ts=2023-03-06T13:09:31.373Z caller=web.go:540 component=web msg="Start listening for connections" address=0.0.0.0:9090
level=info ts=2023-03-06T13:09:31.374Z caller=main.go:803 msg="Starting TSDB ..."
level=info ts=2023-03-06T13:09:31.377Z caller=tls_config.go:191 component=web msg="TLS is disabled." http2=false
level=info ts=2023-03-06T13:09:31.377Z caller=head.go:741 component=tsdb msg="Replaying on-disk memory mappable chunks if any"
level=info ts=2023-03-06T13:09:31.377Z caller=head.go:755 component=tsdb msg="On-disk memory mappable chunks replay completed" duration=2.033µs
level=info ts=2023-03-06T13:09:31.377Z caller=head.go:761 component=tsdb msg="Replaying WAL, this may take a while"
level=info ts=2023-03-06T13:09:31.378Z caller=head.go:813 component=tsdb msg="WAL segment loaded" segment=0 maxSegment=0
level=info ts=2023-03-06T13:09:31.378Z caller=head.go:818 component=tsdb msg="WAL replay completed" checkpoint_replay_duration=16.137µs wal_replay_duration=155.043µs total_replay_duration=184.784µs
level=info ts=2023-03-06T13:09:31.378Z caller=main.go:828 fs_type=EXT4_SUPER_MAGIC
level=info ts=2023-03-06T13:09:31.378Z caller=main.go:831 msg="TSDB started"
level=info ts=2023-03-06T13:09:31.378Z caller=main.go:957 msg="Loading configuration file" filename=/etc/prometheus/config_out/prometheus.yml
level=info ts=2023-03-06T13:09:31.382Z caller=kubernetes.go:266 component="discovery manager scrape" discovery=kubernetes msg="Using pod service account via in-cluster config"
level=info ts=2023-03-06T13:09:31.424Z caller=main.go:988 msg="Completed loading of configuration file" filename=/etc/prometheus/config_out/prometheus.yml totalDuration=45.173445ms remote_storage=1.584µs web_handler=349ns query_engine=647ns scrape=237.087µs scrape_sd=2.202854ms notify=38.233µs notify_sd=29.058µs rules=39.20739ms
level=info ts=2023-03-06T13:09:31.424Z caller=main.go:775 msg="Server is ready to receive web requests."
```

**Comment by roidelapluie** [comment]
I noticed that your UI is running on port 3005 while your api is on port 8082. This is the reason why you are currently getting CORS errors. Can you please confirm if your external URL is set correctly in the Prometheus CLI flags?

**Comment by raj-suvariya** [comment]
@roidelapluie 
Yes, the test setup is UI (our custom UI) is running on localhost:3005 and the Prometheus is running on K8s pod which is port forwarded to localhost:8082. Do we need to set up an external URL for using the Prometheus server's rest APIs? Also, I couldn't find any such documentation. Can you point me to the correct documentation if it exists?

**Comment by roidelapluie** [comment]
We automatically add the Access-Control-Allow-Origin header if there is an Origin header, which is typically handled by the browser. It appears that there may be a bug in your custom UI that is causing the CORS error.

To resolve this issue, you may try serving both the UI and Prometheus with the same reverse proxy to avoid CORS issues. Alternatively, you can investigate the bug in your custom UI that is causing the issue.

**Comment by raj-suvariya** [comment]
@roidelapluie having both UI and Prometheus on the same server is not possible. I will have to look for the actual issue.

Just to get better understanding are you saying that if there is an origin header coming in the request then the server will allow it? Also, is there a way I can find what is happening on server by probably changing the log level?

**Comment by raj-suvariya** [comment, marked as answer]
@roidelapluie this indeed was my mistake.

I was passing `method: "GET"` in the header while making a fetch call from js file. Instead, it is supposed to be in the init options of the fetch call. Thanks for the support!
